x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+		return nil, types.ErrGVGNotExistInFamily
+	}
+
+	var gvgsBindingOnBucket *types.GlobalVirtualGroupsBindingOnBucket
+	var lvg *types.LocalVirtualGroup
+	gvgsBindingOnBucket, found = k.GetGVGsBindingOnBucket(ctx, bucketID)
+	var newLVG = false
+	if !found {
+		// Create a new key store the gvgs binding on bucket
+		lvgID := k.GenNextLVGID(ctx)
+		lvg = &types.LocalVirtualGroup{
+			Id:                   lvgID,
+			GlobalVirtualGroupId: gvgID,
+			StoredSize:           payloadSize,
+			BucketId:             bucketID,
+		}
+		newLVG = true
+		gvgsBindingOnBucket = &types.GlobalVirtualGroupsBindingOnBucket{
+			BucketId: bucketID,
+		}
+		gvgsBindingOnBucket.AppendGVGAndLVG(gvgID, lvgID)
+	} else {
+		lvgID := gvgsBindingOnBucket.GetLVGIDByGVGID(gvgID)
+		if lvgID == 0 {
+			if k.MaxLocalVirtualGroupNumPerBucket(ctx) < uint32(len(gvgsBindingOnBucket.LocalVirtualGroupIds)) {
+				return nil, types.ErrLimitationExceed.Wrapf("The lvg number within the bucket exceeds the limit")
+			}
+			// not exist
+			lvgID = k.GenNextLVGID(ctx)
+			lvg = &types.LocalVirtualGroup{
+				Id:                   lvgID,
+				GlobalVirtualGroupId: gvgID,
+				StoredSize:           payloadSize,
+				BucketId:             bucketID,
+			}
+			newLVG = true
+			gvgsBindingOnBucket.AppendGVGAndLVG(gvgID, lvgID)
+		} else {
+			lvg, found = k.GetLVG(ctx, bucketID, lvgID)
+			if !found {
+				return nil, types.ErrLVGNotExist
+			}
+			lvg.StoredSize += payloadSize
+		}
+	}
+
+	gvg.StoredSize += payloadSize
+
+	k.SetGVG(ctx, gvg)
+	k.SetLVG(ctx, lvg)
+	k.SetGVGsBindingOnBucket(ctx, gvgsBindingOnBucket)
+
+	if newLVG {
+		if err := ctx.EventManager().EmitTypedEvents(&types.EventCreateLocalVirtualGroup{
+			Id:                   lvg.Id,
+			BucketId:             lvg.BucketId,
+			GlobalVirtualGroupId: lvg.GlobalVirtualGroupId,
+			StoredSize:           lvg.StoredSize,
+		}); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateLocalVirtualGroup{
+			Id:                   lvg.Id,
+			GlobalVirtualGroupId: lvg.GlobalVirtualGroupId,
+			StoredSize:           lvg.StoredSize,
+		}); err != nil {
+			return nil, err
+		}
+	}
+	return lvg, nil
+}
+
+func (k Keeper) UnBindingObjectFromLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32, payloadSize uint64) error {
+	lvg, found := k.GetLVG(ctx, bucketID, lvgID)
+	if !found {
+		return types.ErrLVGNotExist
+	}
+	gvgsBindingOnBucket, found := k.GetGVGsBindingOnBucket(ctx, bucketID)
+	if !found {
+		panic(fmt.Sprintf("gvgs binding on bucket mapping not found, bucketID: %s", bucketID.String()))
+	}
+	gvgID := gvgsBindingOnBucket.GetGVGIDByLVGID(lvgID)
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		ctx.Logger().Error("GVG Not Exist, bucketID: %s, gvgID: %d, lvgID :%d", bucketID.String(), gvgID, lvgID)
+		return types.ErrGVGNotExist
+	}
+
+	lvg.StoredSize -= payloadSize
+	gvg.StoredSize -= payloadSize
+
+	k.SetLVG(ctx, lvg)
+	k.SetGVG(ctx, gvg)
```

Comment:
It is not proper to name as `UnBindingObjectFromLVG` if it try to unbind `GVG` as well.